Escape command lines properly

keszybz opened the pull request.

poettering commented on src/basic/escape.h:
```diff
@@ -33,15 +33,12 @@ typedef enum UnescapeFlags {
-        ESCAPE_POSIX             = 3,  /* Similar to ESCAPE_BACKSLASH_ONELINE, but uses POSIX shell escape
-                                        * syntax (a string enclosed in $'') instead of plain quotes. */
-} EscapeStyle;
+typedef enum EscapeFlags {
```
maybe rename ShellEscapeFlags while you touch this? This is after all unrelated to the more generic escape calls such as cescape, but strictly about shell escaping, afaics

poettering commented on src/basic/escape.h:
```diff
@@ -33,15 +33,12 @@ typedef enum UnescapeFlags {
+        /* The default is to add shell quotes ("") so the shell will consider this a single argument.
+         * Tabs and newlines are escaped. */
+
+        ESCAPE_POSIX   = 1 << 1, /* Use POSIX shell escape syntax (a string enclosed in $'') instead of plain quotes. */
```
weird spaces before the =

poettering commented on src/basic/escape.c:
```diff
@@ -510,7 +510,7 @@ char* shell_maybe_quote(const char *s, EscapeFlags flags) {
 
-        if (!*p)
+        if (!*p && (p > s || !FLAGS_SET(flags, ESCAPE_EMPTY)))
                 return strdup(s);
```
i think it would be a lot more readable to place a separate check at the top of the function:

```c
if (FLAGS_SET(flags, ESCAPE_EMPTY) && isempty(s))
        return strdup("\"\"");
```

i.e. this is a special case, lets give it a special codepath...

poettering commented on src/basic/string-util.h:
```diff
@@ -129,6 +129,9 @@ static inline bool _pure_ in_charset(const char *s, const char* charset) {
 }
 
+static inline bool char_is_cc(char p) {
+        return (p > 0 && p < ' ') || p == 127;
```
hmm NUL is a cc too I'd claim

poettering commented on src/basic/escape.c:
```diff
@@ -505,19 +501,18 @@ char* shell_maybe_quote(const char *s, EscapeFlags flags) {
-                if (*p <= ' ' ||
-                    *p >= 127 ||
-                    strchr(SHELL_NEED_QUOTES, *p))
+                if (char_is_cc(*p) ||
```
this changes the test a bit. previously all non-asciii chars were escaped, they are no longer. which is probably  OK, but needs a commit msg comment at least

poettering commented on src/basic/process-util.c:
```diff
@@ -158,51 +165,103 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
+                        if (!arg2)
+                                return -ENOMEM;
+
+                        ans = shell_maybe_quote(arg2, 0);
```
we almost always prefer C style escaping over shell encoding. Shouldn't we continue to stick to that, and only use shell style escaping when we really know we want it?

or in other words: make the escaping style configured via a flag? I mean, there's already a flags param, might as well extend that?

poettering commented on src/basic/process-util.c:
```diff
@@ -158,51 +165,103 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
+                        if (!escaped)
+                                return -ENOMEM;
+
+                        r = strv_consume(&args, escaped ? TAKE_PTR(escaped) : TAKE_PTR(arg));
```
maybe:

```c
TAKE_PTR(escaped) ?: TAKE_PTR(arg)
```

poettering reviewed (changes requested):
looks pretty good, some comments

keszybz commented:
Updated with all the comments marked as resolved resolved.

Your comment about c escaping got me thinking, and I added one more commit that actually goes in the opposite direction: it uses posix-style $'' escapes in coredump. See the last commit for justification.

mrc0mmand commented:
```
==8914==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602000000d79 at pc 0x7fd3c3fcd9c6 bp 0x7fff69bc6130 sp 0x7fff69bc6128
WRITE of size 1 at 0x602000000d79 thread T0 (aaaa bbbb cccc)
    #0 0x7fd3c3fcd9c5 in cescape_char /systemd-meson-build/../build/src/basic/escape.c:66:42
    #1 0x7fd3c3fd0b27 in strcpy_backslash_escaped /systemd-meson-build/../build/src/basic/escape.c:470:30
    #2 0x7fd3c3fd0fce in shell_maybe_quote /systemd-meson-build/../build/src/basic/escape.c:525:13
    #3 0x7fd3c404ae67 in get_process_cmdline /systemd-meson-build/../build/src/basic/process-util.c:233:35
    #4 0x5633f35167b4 in test_get_process_cmdline_harder /systemd-meson-build/../build/src/test/test-process-util.c:416:9
    #5 0x5633f351085c in main /systemd-meson-build/../build/src/test/test-process-util.c:749:9
    #6 0x7fd3c369fb24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #7 0x5633f35105dd in _start (/systemd-meson-build/test-process-util+0xe5dd)

0x602000000d79 is located 0 bytes to the right of 9-byte region [0x602000000d70,0x602000000d79)
allocated by thread T0 (aaaa bbbb cccc) here:
    #0 0x7fd3c4932009 in malloc (/usr/lib/clang/11.1.0/lib/linux/libclang_rt.asan-x86_64.so+0xf2009)
    #1 0x7fd3c3fcde74 in malloc_multiply /systemd-meson-build/../build/src/basic/alloc-util.h:96:16
    #2 0x7fd3c3fd0ef4 in shell_maybe_quote /systemd-meson-build/../build/src/basic/escape.c:512:15
    #3 0x7fd3c404ae67 in get_process_cmdline /systemd-meson-build/../build/src/basic/process-util.c:233:35
    #4 0x5633f35167b4 in test_get_process_cmdline_harder /systemd-meson-build/../build/src/test/test-process-util.c:416:9
    #5 0x5633f351085c in main /systemd-meson-build/../build/src/test/test-process-util.c:749:9
    #6 0x7fd3c369fb24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

SUMMARY: AddressSanitizer: heap-buffer-overflow /systemd-meson-build/../build/src/basic/escape.c:66:42 in cescape_char
```

(Also, rebasing against current master should make Packit green)

keszybz commented:
Thanks @mrc0mmand. The bug should be fixed now.

Anyway, since this changes behaviour quite a bit, I don't think it should be merged before the release.

poettering commented on src/basic/escape.c:
```diff
@@ -462,11 +462,11 @@ char* octescape(const char *s, size_t len) {
 
```
i figure for the commit mgs title s/also/always/?

poettering commented on src/basic/process-util.c:
```diff
@@ -136,12 +136,19 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
+         *   output. Non-utf8 bytes are replaced by �. The returned string is of the specified console width
+         *   at most, abbreviated with an ellipsis.
+         * - otherwise, the output is shell-quoted and escaped, and in principle can by copy-pasted into
+         *   the terminal to execute. UTF-8 output is assumed.
```
it sounds so strange that thus function has no way to retrieve the argv[] array 1:1 like it would be passed to main(). I have the suspicion we should hide the shell escaping logic behind some flag, and if the flag is not specified return the unmodified argv[] basically. 

We might not use it that way in our regular codepaths, but I'd still do it, and it might even be useful in tests.

Alternatively, rename the function to get_process_cmdline_shell_escaped() or so, to make very clear the escaped cmdline is the only one you'll get

poettering commented on src/basic/process-util.c:
```diff
@@ -158,51 +165,103 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
         if ((size_t) 4 * max_columns + 1 < max_columns)
```
this overflow check is icky btw

poettering commented on src/basic/process-util.c:
```diff
@@ -158,51 +165,103 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
+                        if (!escaped)
+                                return -ENOMEM;
+
+                        r = strv_consume(&args, TAKE_PTR(escaped) ?: TAKE_PTR(arg));
```
you checked that escaped is non-NULL the line before, so the ternary op doesn't make much sense?

poettering reviewed (changes requested):
looks pretty good. just some minor thing

keszybz commented:
I'll wait with rebasing this until #19085 goes in, since it touches the same cod.

keszybz commented:
I reworked this quite a bit, a fresh review will be required :(
All the formatting code and truncation with utf8 and non-utf8 gave me a lot of grief, but I think it's all working fine now, even better than before.

mrc0mmand commented:
LSan is not happy:

```
==46==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 6 byte(s) in 6 object(s) allocated from:
    #0 0x7f5133e7e392 in realloc (/usr/lib/clang/11.1.0/lib/linux/libclang_rt.asan-x86_64.so+0xf2392)
    #1 0x7f513352b1ae in read_virtual_file /systemd-meson-build/../build/src/basic/fileio.c:471:21
    #2 0x7f5133598527 in get_process_cmdline_nulstr /systemd-meson-build/../build/src/basic/process-util.c:144:13
    #3 0x7f51335979f8 in get_process_cmdline /systemd-meson-build/../build/src/basic/process-util.c:202:13
    #4 0x5608aa743dca in client_context_read_basic /systemd-meson-build/../build/src/journal/journald-context.c:228:13
    #5 0x5608aa740a91 in client_context_really_refresh /systemd-meson-build/../build/src/journal/journald-context.c:519:9
    #6 0x5608aa742733 in client_context_get_internal /systemd-meson-build/../build/src/journal/journald-context.c:705:9
    #7 0x5608aa7421ed in client_context_get /systemd-meson-build/../build/src/journal/journald-context.c:719:16
    #8 0x5608aa7491ef in server_process_native_message /systemd-meson-build/../build/src/journal/journald-native.c:310:21
    #9 0x5608aa70fd96 in server_process_datagram /systemd-meson-build/../build/src/journal/journald-server.c:1354:25
    #10 0x7f51338625bc in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:3512:21
    #11 0x7f5133861570 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4077:21
    #12 0x7f513386428c in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4138:21
    #13 0x5608aa6f5ad8 in main /systemd-meson-build/../build/src/journal/journald.c:107:21
    #14 0x7f5132a9fb24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f5133e7e392 in realloc (/usr/lib/clang/11.1.0/lib/linux/libclang_rt.asan-x86_64.so+0xf2392)
    #1 0x7f513352b1ae in read_virtual_file /systemd-meson-build/../build/src/basic/fileio.c:471:21
    #2 0x7f5133598527 in get_process_cmdline_nulstr /systemd-meson-build/../build/src/basic/process-util.c:144:13
    #3 0x7f51335979f8 in get_process_cmdline /systemd-meson-build/../build/src/basic/process-util.c:202:13
    #4 0x5608aa743dca in client_context_read_basic /systemd-meson-build/../build/src/journal/journald-context.c:228:13
    #5 0x5608aa740a91 in client_context_really_refresh /systemd-meson-build/../build/src/journal/journald-context.c:519:9
    #6 0x5608aa742733 in client_context_get_internal /systemd-meson-build/../build/src/journal/journald-context.c:705:9
    #7 0x5608aa742cdd in client_context_acquire /systemd-meson-build/../build/src/journal/journald-context.c:730:16
    #8 0x5608aa72de7d in stdout_stream_log /systemd-meson-build/../build/src/journal/journald-stream.c:266:21
    #9 0x5608aa72b62d in stdout_stream_line /systemd-meson-build/../build/src/journal/journald-stream.c:444:24
    #10 0x5608aa72b192 in stdout_stream_found /systemd-meson-build/../build/src/journal/journald-stream.c:465:13
    #11 0x5608aa72aecd in stdout_stream_scan /systemd-meson-build/../build/src/journal/journald-stream.c:509:21
    #12 0x5608aa7267fe in stdout_stream_process /systemd-meson-build/../build/src/journal/journald-stream.c:608:13
    #13 0x7f51338625bc in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:3512:21
    #14 0x7f5133861570 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4077:21
    #15 0x7f513386428c in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4138:21
    #16 0x5608aa6f5ad8 in main /systemd-meson-build/../build/src/journal/journald.c:107:21
    #17 0x7f5132a9fb24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

SUMMARY: AddressSanitizer: 7 byte(s) leaked in 7 allocation(s).
```

keszybz commented:
> LSan is not happy:

Thanks. I think I found the issue.

mrc0mmand commented:
> > LSan is not happy:
> 
> Thanks. I think I found the issue.

Almost there!

```
==52574==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 4240 byte(s) in 178 object(s) allocated from:
    #0 0x7f888e752009 in malloc (/usr/lib/clang/11.1.0/lib/linux/libclang_rt.asan-x86_64.so+0xf2009)
    #1 0x7f888de9f8de in malloc_multiply /systemd-meson-build/../build/src/basic/alloc-util.h:96:16
    #2 0x7f888dea253d in strv_join_full /systemd-meson-build/../build/src/basic/strv.c:379:13
    #3 0x7f888de69b33 in strv_join /systemd-meson-build/../build/src/basic/strv.h:103:16
    #4 0x7f888de68ed8 in get_process_cmdline /systemd-meson-build/../build/src/basic/process-util.c:229:23
    #5 0x5644bdee7df3 in test_get_process_cmdline_one /systemd-meson-build/../build/src/test/test-process-util.c:108:13
    #6 0x5644bdedf147 in test_get_process_cmdline /systemd-meson-build/../build/src/test/test-process-util.c:138:17
    #7 0x5644bdedd8e7 in main /systemd-meson-build/../build/src/test/test-process-util.c:860:9
    #8 0x7f888d4b7b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Direct leak of 1544 byte(s) in 33 object(s) allocated from:
    #0 0x7f888e752009 in malloc (/usr/lib/clang/11.1.0/lib/linux/libclang_rt.asan-x86_64.so+0xf2009)
    #1 0x7f888de9f8de in malloc_multiply /systemd-meson-build/../build/src/basic/alloc-util.h:96:16
    #2 0x7f888dea253d in strv_join_full /systemd-meson-build/../build/src/basic/strv.c:379:13
    #3 0x7f888de69b33 in strv_join /systemd-meson-build/../build/src/basic/strv.h:103:16
    #4 0x7f888de68ed8 in get_process_cmdline /systemd-meson-build/../build/src/basic/process-util.c:229:23
    #5 0x5644bdee7d3f in test_get_process_cmdline_one /systemd-meson-build/../build/src/test/test-process-util.c:105:13
    #6 0x5644bdedf147 in test_get_process_cmdline /systemd-meson-build/../build/src/test/test-process-util.c:138:17
    #7 0x5644bdedd8e7 in main /systemd-meson-build/../build/src/test/test-process-util.c:860:9
    #8 0x7f888d4b7b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

SUMMARY: AddressSanitizer: 5784 byte(s) leaked in 211 allocation(s).
```

keszybz commented:
bionic-i386 seems to have died on the way:
autopkgtest [15:17:45]: ERROR: "mkdir -p /tmp/autopkgtest.0i93id" failed with stderr "Connection to 10.220.42.214 closed by remote host."

bionic-arm64: `375/552 test-random-util                        TIMEOUT 30.01 s `

keszybz commented:
exec_command_dump() in src/core/ should be converted to use this too.

poettering commented on src/basic/escape.c:
```diff
@@ -504,6 +504,9 @@ char* shell_maybe_quote(const char *s, ShellEscapeFlags flags) {
          * escaping too, but that should be OK. */
 
+        if (FLAGS_SET(flags, SHELL_ESCAPE_EMPTY) && isempty(s))
+                return strdup("''"); /* We don't use $'' here in the POSIX mode. '' is fine too. */
```
hmm, it's a bit weird to return `''` instead of `""` here, no? i mean, it sounds more systematic if we try to cut down on different escape modes, i.e. I'd really return `""` if posix mode is off here, and otherwise `$''`, but at the very least use `""` instead of `''`. but returning `""` if the user explicitly requested posix escaping via the flag sounds weird too

the commit msg is not inline with the used quotes here btw, it talks partially about `""` and in the title about `''`...

keszybz replied:
I changed it to return "" now.

poettering commented on src/basic/string-util.h:
```diff
@@ -129,6 +129,9 @@ static inline bool _pure_ in_charset(const char *s, const char* charset) {
 }
 
+static inline bool char_is_cc(char p) {
+        return (p >= 0 && p < ' ') || p == 127;
```
hmm, on archs that have unsigned chars the first check might result in a warning as being redundant... maybe cast to unsigned here explicitly...

(but also, if the CI is happy I don't care)

keszybz replied:
I don't have access to a BE machine. But I think gcc/clang are smart enough to know that `char` is special…

poettering commented on src/basic/fileio.c:
```diff
@@ -398,21 +404,21 @@ int read_full_virtual_file(const char *filename, char **ret_contents, size_t *re
                 assert_cc(READ_FULL_BYTES_MAX < SSIZE_MAX);
                 if (st.st_size > 0) {
-                        if (st.st_size > READ_FULL_BYTES_MAX)
+                        size = MIN((size_t) st.st_size, max_size);
```
the could possibly overflow on 32bit userspace on a 64bit kernel (i.e you run this on /proc/kcore), because the file size is going to be 64bit, but size_t is 32bit.

poettering commented on src/basic/fileio.c:
```diff
@@ -363,32 +363,38 @@ int verify_file(const char *fn, const char *blob, bool accept_extra_nl) {
 }
 
-int read_full_virtual_file(const char *filename, char **ret_contents, size_t *ret_size) {
+int read_virtual_file(const char *filename, size_t max_size, char **ret_contents, size_t *ret_size) {
```
wouldn't read_full_virtual_file_full() be  a better name?

keszybz replied:
The patch makes it *not* read the full file… So I think it's fine to call it `read_virtual_file` without further details in the name.

poettering commented on src/basic/escape.c:
```diff
@@ -360,15 +360,16 @@ int cunescape_length_with_prefix(const char *s, size_t length, const char *prefi
 }
 
-char* xescape_full(const char *s, const char *bad, size_t console_width, bool eight_bits) {
+char* xescape_full(const char *s, const char *bad, size_t console_width, bool eight_bits, bool force_truncated) {
```
flagsify?

poettering replied:
als, maybe call this "force_ellipsis"?

poettering commented on src/basic/process-util.c:
```diff
@@ -175,7 +175,7 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
         bool eight_bit = (flags & PROCESS_CMDLINE_USE_LOCALE) && !is_locale_utf8();
 
-        ans = escape_non_printable_full(t, max_columns, eight_bit);
+        ans = escape_non_printable_full(t, max_columns, eight_bit, false);
```
maybe add `/* force_ellispsis= */` before the `false`

(or better: turn this into a flag)

poettering commented on src/basic/process-util.c:
```diff
@@ -123,64 +123,128 @@ int get_process_comm(pid_t pid, char **ret) {
 
-int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags, char **line) {
-        _cleanup_free_ char *t = NULL, *ans = NULL;
+static int get_process_cmdline_nulstr(pid_t pid, size_t max_size, ProcessCmdlineFlags flags, char **ret, size_t *ret_size) {
```
maybe line break args, got so long...

poettering commented on src/basic/process-util.c:
```diff
@@ -123,64 +123,128 @@ int get_process_comm(pid_t pid, char **ret) {
+        r = get_process_cmdline_nulstr(pid, max_columns, flags, &t, &k);
+        if (r < 0)
+                return r;
+        bool truncated = r == 0;
```
i'd use `int` a type for `truncated` and then do:

```c
truncated = get_process_cmdline_nulstr();
if (truncated < 0)
        return truncated;
```

without the extra step through `r`, but doesn't matter

keszybz replied:
OK.

poettering commented on src/basic/process-util.c:
```diff
@@ -123,64 +123,128 @@ int get_process_comm(pid_t pid, char **ret) {
+                if (!args)
+                        return -ENOMEM;
+
+                for (size_t i = 0; args[i]; i++) {
```
STRV_FOREACH?

poettering commented on src/test/test-process-util.c:
```diff
@@ -90,6 +92,53 @@ static void test_get_process_comm(pid_t pid) {
+
+        assert_se(d = opendir("/proc"));
+
+
```
double empty line

keszybz commented:
Updated with the requested changes (except as commented above).

mrc0mmand commented:
> From https://man7.org/linux/man-pages/man5/proc.5.html it sounds like the correct way of interpreting double null bytes is as a terminator, so systemd needs to be fixed.
> 
> ```
>        /proc/[pid]/cmdline
>               This read-only file holds the complete command line for
>               the process, unless the process is a zombie.  In the
>               latter case, there is nothing in this file: that is, a
>               read on this file will return 0 characters.  The command-
>               line arguments appear in this file as a set of strings
>               separated by null bytes ('\0'), with a further null byte
>               after the last string.
> ```

Could you, please, open a new issue, so this won't get lost?

bobrik commented:
Sure, filed: https://github.com/systemd/systemd/issues/21186.
